src/java.base/share/classes/java/security/PEMDecoder.java

```diff
@@ -0,0 +1,411 @@
+                        X509EncodedKeySpec spec = new X509EncodedKeySpec(
+                            p8.getPubKeyEncoded(), algo);
+                        yield new KeyPair(getKeyFactory(algo).
+                            generatePublic(spec), p8);
+                    } else {
+                        // No public key, return the private key.
+                        yield d;
+                    }
+                }
+                case Pem.ENCRYPTED_PRIVATE_KEY -> {
+                    if (password == null) {
+                        yield new EncryptedPrivateKeyInfo(decoder.decode(
+                            pem.pem()));
+                    }
+                    yield new EncryptedPrivateKeyInfo(decoder.decode(pem.pem())).
+                        getKey(password.getPassword());
+                }
+                case Pem.CERTIFICATE, Pem.X509_CERTIFICATE -> {
+                    CertificateFactory cf = getCertFactory("X509");
+                    yield (X509Certificate) cf.generateCertificate(
+                        new ByteArrayInputStream(decoder.decode(pem.pem())));
+                }
+                case Pem.X509_CRL -> {
+                    CertificateFactory cf = getCertFactory("X509");
+                    yield (X509CRL) cf.generateCRL(
+                        new ByteArrayInputStream(decoder.decode(pem.pem())));
+                }
+                case Pem.RSA_PRIVATE_KEY -> {
+                    KeyFactory kf = getKeyFactory("RSA");
+                    yield kf.generatePrivate(
+                        RSAPrivateCrtKeyImpl.getKeySpec(decoder.decode(
+                            pem.pem())));
+                }
+                default -> pem;
+            };
+        } catch (GeneralSecurityException | IOException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Decodes and returns a {@link DEREncodable} from the given string.
+     *
+     * @param str a String containing PEM data.
+     * @return a {@code DEREncodable} generated from the PEM data.
+     * @throws IllegalArgumentException on error in decoding or if the PEM is
+     * unsupported.
+     * @throws NullPointerException when {@code str} is null.
+     */
+    public DEREncodable decode(String str) {
+        Objects.requireNonNull(str);
+        try {
+            return decode(new ByteArrayInputStream(
+                str.getBytes(StandardCharsets.ISO_8859_1)));
+        } catch (IOException e) {
+            // With all data contained in the String, there are no IO ops.
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Decodes and returns a {@link DEREncodable} from the given
+     * {@code InputStream}.
+     *
+     * <p>This method will read the {@code InputStream} until PEM data is
+     * found or until the end of the stream.  Non-PEM data in the
+     * {@code InputStream} before the PEM header will be ignored by the decoder.
```

Comment:
1. This is not precise because you might put the leading data into a `PEMRecord`.

Also, I think it's very important to point out the pointer position of the input stream after this method is successfully called. Otherwise, the method cannot be reliably used in extracting multiple PEMs from a single stream.

Also, I'm not exactly sure what happens when there is nothing more to read.

Reply:
See https://github.com/openjdk/jdk/blob/002679ac9fe4de8150b7dd4c9aeb44eeef1257d6/src/java.base/share/classes/java/security/cert/CertificateFactory.java#L338-L347